tests/diff/parse.c

```diff
@@ -288,3 +288,74 @@ void test_diff_parse__patch_roundtrip_succeeds(void)
 	git_buf_dispose(&patchbuf);
 	git_buf_dispose(&diffbuf);
 }
+
+#define cl_assert_equal_i_src(i1,i2,file,line) clar__assert_equal(file,line,#i1 " != " #i2, 1, "%d", (int)(i1), (int)(i2))
+
+static void cl_git_assert_lineinfo_(int old_lineno, int new_lineno, int num_lines, git_patch *patch, size_t hunk_idx, size_t line_idx, const char *file, int lineno)
+{
+	const git_diff_line *line;
+
+	cl_git_expect(git_patch_get_line_in_hunk(&line, patch, hunk_idx, line_idx), 0, file, lineno);
+	cl_assert_equal_i_src(old_lineno, line->old_lineno, file, lineno);
+	cl_assert_equal_i_src(new_lineno, line->new_lineno, file, lineno);
+	cl_assert_equal_i_src(num_lines, line->num_lines, file, lineno);
+}
+
+#define cl_git_assert_lineinfo(old, new, num, p, h, l) \
+	cl_git_assert_lineinfo_(old,new,num,p,h,l,__FILE__,__LINE__)
+
+
+void test_diff_parse__issue4672(void)
+{
+	const char *text = "diff --git a/a b/a\n"
+	"index 7f129fd..af431f2 100644\n"
+	"--- a/a\n"
+	"+++ b/a\n"
+	"@@ -3 +3 @@\n"
+	"-a contents 2\n"
+	"+a contents\n";
+
+	git_diff *diff;
+	git_patch *patch;
+	const git_diff_hunk *hunk;
+	size_t n, l = 0;
+
+	cl_git_pass(git_diff_from_buffer(&diff, text, strlen(text)));
+	cl_git_pass(git_patch_from_diff(&patch, diff, 0));
+	cl_git_pass(git_patch_get_hunk(&hunk, &n, patch, 0));
+
+	cl_git_assert_lineinfo(3, -1, 1, patch, 0, l++);
+	cl_git_assert_lineinfo(-1, 3, 1, patch, 0, l++);
+
+	cl_assert_equal_i(n, l);
+
+	git_patch_free(patch);
+	git_diff_free(diff);
+}
+
+void test_diff_parse__lineinfo(void)
+{
+	const char *text = PATCH_ORIGINAL_TO_CHANGE_MIDDLE;
+	git_diff *diff;
+	git_patch *patch;
+	const git_diff_hunk *hunk;
+	size_t n, l = 0;
+
+	cl_git_pass(git_diff_from_buffer(&diff, text, strlen(text)));
+	cl_git_pass(git_patch_from_diff(&patch, diff, 0));
+	cl_git_pass(git_patch_get_hunk(&hunk, &n, patch, 0));
+
+	cl_git_assert_lineinfo(3, 3, 1, patch, 0, l++);
+	cl_git_assert_lineinfo(4, 4, 1, patch, 0, l++);
+	cl_git_assert_lineinfo(5, 5, 1, patch, 0, l++);
+	cl_git_assert_lineinfo(6, -1, 1, patch, 0, l++);
+	cl_git_assert_lineinfo(-1, 6, 1, patch, 0, l++);
+	cl_git_assert_lineinfo(7, 7, 1, patch, 0, l++);
+	cl_git_assert_lineinfo(8, 8, 1, patch, 0, l++);
+	cl_git_assert_lineinfo(9, 9, 1, patch, 0, l++);
```

Comment:
I don't love this.  I realize that you're aiming for brevity here but it's _really_ hard for me to read and understand what this at a glance.  I would much rather you unrolled these into simple tests.

Reply:
And it's not even complete… Here's an [example](https://github.com/libgit2/libgit2/blob/master/tests/diff/patch.c?utf8=%E2%9C%93#L313) of the expanded version. My one isn't as exhaustive (origin and offset are missing), but it tests each line, so it's about 5 lines of asserts * 9 lines of patch => 45 lines of test.

Would something like this be more acceptable ?
```
cl_git_pass(git_line_from_hunk_patch_diff(&line, patch, 0, l++));
cl_git_assert_lineinfo(line, origin, old, new, newlines, offset, "stuff\n");
```

Reply:
Don't use an expanded version, but simply use an array :)

```
struct {
    size_t start;
    size_t end;
    size_t length; /* No idea what those fields really are, but you get what I mean */
} expected_lineinfos[] = {
    { 3, 3, 1 },
    { 4, 4, 1 },
    ...
};

for (i = 0; i < ARRAY_SIZE(expected_lineinfos); i++) {
    git_line_from_hunk_patch_diff(&line, patch, 0, i);
    cl_assert_equal_i(old_lineno, line->old_lineno);
    cl_assert_equal_i(new_lineno, line->new_lineno);
    cl_assert_equal_i(num_lines, line->num_lines);
}
```

Reply:
⚠️ Incoming bikeshedding.

I'm not against going the array route (I contemplated doing that). What makes me prefer the macro route is that I can preserve "intent" (eg. `cl_git_assert_lineinfo` vs a bunch of `equal_i`), as well as `__LINE__` numbers, the latter helps to make the failure context relevant straight from CI (which is something I'm personally striving for when writing tests (hence my propension to add macros on top of helper functions). There are quite many tests that check line info around, and that, as pointed out above, more would be quite welcome (eg. I haven't found anything that test hunk headers context lines behavior), and thus my preference goes to a tight one-liner.

The array route doesn't allow for that preservation, and even though it looks cleaner, it's messier to preserve the line info with them. I have a feeling I have already wrote one of these array-ed+macro tests somewhere, which, given my constraints, was 😒.

Reply:
I'm torn.  I appreciate that you're trying to report line numbers, though it's not my highest priority.  I'd rather have readable tests that convey intent than to know the line number.  eg:

```
cl_assert_equal_i(old_lineno, line->old_lineno);
cl_assert_equal_i(new_lineno, line->new_lineno);
cl_assert_equal_i(num_lines, line->num_lines);
```

makes it _incredibly_ clear what we're doing when I look at this.  Now, I concede that regressions would be a bit challenging -- we would have an accurate line number, but that wouldn't necessarily correspond to the failing test data.  We would know, for example, that the `new_lineno` test failed, but not what the data was that caused it to fail, since it could have been any data from the array.

This is a pattern we have in a lot of places and I don't think that we see a lot of regressions around them.  So I'm inclined to just shrug and say it's okay.  I definitely _have_ done big refactorings where this was a bit annoying - for example, when refactoring the iterators - but it didn't scar me so much that I want to avoid it.

Is your motivation borne out of past experiences with this pattern?  Or are you trying to future proof things?  Again, my suggestion is that we go for simplicity in tests over strict failure-reporting accuracy here, but we can probably spend some time working to find a more readable compromise if you disagree.

Reply:
Well, in my opinion it doesn't help much to have a one-shot test helper that's only ever going to be used once. In that case, you can just as well inline it, as you'd be using line info anyway. I wanted to propose to just use `cl_assert_equal_i_fmt`, but it turns out `fmt` is not really a format but simply a string where the two ints should be printed into. Would've been nice to have a macro `cl_assert_equal_i_fmt(tests[i].old, line->old_lineno, "old line numbers not matching: expected %i, got %i in line '%s'", tests[i].old, line->old_lineno, line->str)`. But yeah.

Dunno. I'll retreat from this discussion as I do not think it's important enough. Both styles have their advantages and disadvantages.

Reply:
> a one-shot test helper that's only ever going to be used once.

Actually, the plan was to have one helper that checks `line_info` and use that everywhere. That would reinforce the current "generated patchs" test as well (that's what I was hinting at above, since some checks are missing).

As an aside, sometimes I feel limited by clar's architecture, and long for something like `rspec` behaviors… That would allow reusing complete test suites against different "initial conditions" (think bare repo vs non-bare repo worktrees).

Reply:
Yeah, I had the same feeling sometimes about clar. It' works reasonably well and is a breeze to use in most cases, but some kind of data/configuration/permutation driven interface would in some cases be really helpful. I mean you can easily set that up already, but you do loose information then like we see here in this example.